### ISSUE TITLE
"podman generate systemd" for rootless pod

### ISSUE THREAD

**Issue by barnscott** [containers/podman#7330]
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

I am seeing errors when trying to start pod created from "podman generate systemd". i do not have any issues when using systemd unit file that were generated for rootless stand-alone containers, only seeing issue for my rootless pod.

here is the pod im seeing issues with:
```
$ podman ps --pod
CONTAINER ID  IMAGE                                             COMMAND               CREATED       STATUS                PORTS                   NAMES               POD ID        PODNAME
08d7b92c717a  docker.io/library/postgres:latest                 postgres              7 weeks ago   Up 2 seconds ago      0.0.0.0:9088->80/tcp    nc_db               870de8d5fa51  nextcloud
4acaf8eb2e8f  k8s.gcr.io/pause:3.2                                                    7 weeks ago   Up 2 seconds ago      0.0.0.0:9088->80/tcp    870de8d5fa51-infra  870de8d5fa51  nextcloud
4c99d404a174  docker.io/library/nextcloud:latest                apache2-foregroun...  7 weeks ago   Up 2 seconds ago      0.0.0.0:9088->80/tcp    nc_app              870de8d5fa51  nextcloud
```

this is what i am doing to generate unit file:
```
podman generate systemd --name nextcloud > $HOME/.config/systemd/user/nextcloud.service && \
systemctl --user daemon-reload && \
systemctl --user enable --now nextcloud.service 
```

this is the output im getting from above command:
```
Failed to start nextcloud.service: Unit nextcloud.service has a bad unit file setting.
See user logs and 'systemctl --user status nextcloud.service' for details.

```

here is error output:
```
$ systemctl --user status nextcloud.service
● nextcloud.service - Podman container-nc_db.service
     Loaded: bad-setting (Reason: Unit nextcloud.service has a bad unit file setting.)
     Active: failed (Result: exit-code) since Sun 2020-08-16 13:14:11 CDT; 22min ago
       Docs: man:podman-generate-systemd(1)
             man:podman-generate-systemd(1)
             man:podman-generate-systemd(1)
        CPU: 132ms

Aug 16 13:14:11 barnpc systemd[1091]: nextcloud.service: Scheduled restart job, restart counter is at 5.
Aug 16 13:14:11 barnpc systemd[1091]: Stopped Podman pod-nextcloud.service.
Aug 16 13:14:11 barnpc systemd[1091]: nextcloud.service: Start request repeated too quickly.
Aug 16 13:14:11 barnpc systemd[1091]: nextcloud.service: Failed with result 'exit-code'.
Aug 16 13:14:11 barnpc systemd[1091]: Failed to start Podman pod-nextcloud.service.
Aug 16 13:29:58 barnpc systemd[1091]: nextcloud.service: Service has more than one ExecStart= setting, which is only allowed for Type=oneshot services. Refusing.
Aug 16 13:29:58 barnpc systemd[1091]: nextcloud.service: Service has more than one ExecStart= setting, which is only allowed for Type=oneshot services. Refusing.
Aug 16 13:35:54 barnpc systemd[1091]: nextcloud.service: Service has more than one ExecStart= setting, which is only allowed for Type=oneshot services. Refusing.
Aug 16 13:35:54 barnpc systemd[1091]: nextcloud.service: Service has more than one ExecStart= setting, which is only allowed for Type=oneshot services. Refusing.

```

This is the unit file that was created from podman-generate-systemd:
```
$ podman generate systemd --name nextcloud
# pod-nextcloud.service
# autogenerated by Podman 2.0.4
# Sun Aug 16 13:39:55 CDT 2020

[Unit]
Description=Podman pod-nextcloud.service
Documentation=man:podman-generate-systemd(1)
Wants=network.target
After=network-online.target
Requires=container-nc_app.service container-nc_db.service
Before=container-nc_app.service container-nc_db.service

[Service]
Environment=PODMAN_SYSTEMD_UNIT=%n
Restart=on-failure
ExecStart=/usr/bin/podman start 870de8d5fa51-infra
ExecStop=/usr/bin/podman stop -t 10 870de8d5fa51-infra
ExecStopPost=/usr/bin/podman stop -t 10 870de8d5fa51-infra
PIDFile=/tmp/1000/overlay-containers/4acaf8eb2e8f02ed101d764ff267c28275d9e6d61fad3180eadf18de3d005bbc/userdata/conmon.pid
KillMode=none
Type=forking

[Install]
WantedBy=multi-user.target default.target
# container-nc_app.service
# autogenerated by Podman 2.0.4
# Sun Aug 16 13:39:55 CDT 2020

[Unit]
Description=Podman container-nc_app.service
Documentation=man:podman-generate-systemd(1)
Wants=network.target
After=network-online.target
BindsTo=pod-nextcloud.service
After=pod-nextcloud.service

[Service]
Environment=PODMAN_SYSTEMD_UNIT=%n
Restart=on-failure
ExecStart=/usr/bin/podman start nc_app
ExecStop=/usr/bin/podman stop -t 10 nc_app
ExecStopPost=/usr/bin/podman stop -t 10 nc_app
PIDFile=/tmp/1000/overlay-containers/4c99d404a17442069d0008de1e3c20e3916d3a281599d95cfcb417e6d8f70f34/userdata/conmon.pid
KillMode=none
Type=forking

[Install]
WantedBy=multi-user.target default.target
# container-nc_db.service
# autogenerated by Podman 2.0.4
# Sun Aug 16 13:39:55 CDT 2020

[Unit]
Description=Podman container-nc_db.service
Documentation=man:podman-generate-systemd(1)
Wants=network.target
After=network-online.target
BindsTo=pod-nextcloud.service
After=pod-nextcloud.service

[Service]
Environment=PODMAN_SYSTEMD_UNIT=%n
Restart=on-failure
ExecStart=/usr/bin/podman start nc_db
ExecStop=/usr/bin/podman stop -t 10 nc_db
ExecStopPost=/usr/bin/podman stop -t 10 nc_db
PIDFile=/tmp/1000/overlay-containers/08d7b92c717a3b8b8fab9c78ec7d786ce35d638e1240d9623bf840d834d3bcec/userdata/conmon.pid
KillMode=none
Type=forking

[Install]
WantedBy=multi-user.target default.target

```

**Describe the results you received:**
I am see the error:
_nextcloud.service: Service has more than one ExecStart= setting, which is only allowed for Type=oneshot services. Refusing._
I need clarity on the correct way to generate unit file for pod, or a workaround.

**Describe the results you expected:**
I expect it to work similar to how it does with rootless containers (which is what man page describes).

**Additional information you deem important (e.g. issue happens only occasionally):**
NA.
**Output of `podman version`:**
```
$ podman --version
podman version 2.0.4

```

**Output of `podman info --debug`:**

```
$ podman info --debug
host:
  arch: amd64
  buildahVersion: 1.15.0
  cgroupVersion: v2
  conmon:
    package: conmon-2.0.19-1.fc32.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.0.19, commit: 5dce9767526ed27f177a8fa3f281889ad509fea7'
  cpus: 8
  distribution:
    distribution: fedora
    version: "32"
  eventLogger: file
  hostname: barnpc
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
  kernel: 5.7.11-200.fc32.x86_64
  linkmode: dynamic
  memFree: 701116416
  memTotal: 16739414016
  ociRuntime:
    name: crun
    package: crun-0.14.1-1.fc32.x86_64
    path: /usr/bin/crun
    version: |-
      crun version 0.14.1
      commit: 598ea5e192ca12d4f6378217d3ab1415efeddefa
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +YAJL
  os: linux
  remoteSocket:
    path: /run/user/1000/podman/podman.sock
  rootless: true
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: slirp4netns-1.1.4-1.fc32.x86_64
    version: |-
      slirp4netns version 1.1.4
      commit: b66ffa8e262507e37fca689822d23430f3357fe8
      libslirp: 4.3.1
      SLIRP_CONFIG_VERSION_MAX: 2
  swapFree: 34359734272
  swapTotal: 34359734272
  uptime: 1h 26m 44.99s (Approximately 0.04 days)
registries:
  search:
  - registry.fedoraproject.org
  - registry.access.redhat.com
  - registry.centos.org
  - docker.io
store:
  configFile: /var/home/barn/.config/containers/storage.conf
  containerStore:
    number: 7
    paused: 0
    running: 6
    stopped: 1
  graphDriverName: overlay
  graphOptions:
    overlay.mount_program:
      Executable: /usr/bin/fuse-overlayfs
      Package: fuse-overlayfs-1.1.2-1.fc32.x86_64
      Version: |-
        fusermount3 version: 3.9.1
        fuse-overlayfs: version 1.1.0
        FUSE library version 3.9.1
        using FUSE kernel interface version 7.31
  graphRoot: /var/home/barn/.local/share/containers/storage
  graphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "false"
  imageStore:
    number: 23
  runRoot: /tmp/1000
  volumePath: /var/home/barn/.local/share/containers/storage/volumes
version:
  APIVersion: 1
  Built: 0
  BuiltTime: Wed Dec 31 18:00:00 1969
  GitCommit: ""
  GoVersion: go1.14.6
  OsArch: linux/amd64
  Version: 2.0.4
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
$ rpm -q podman
podman-2.0.4-1.fc32.x86_64
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Physical / Fedora Silverblue 32 / rpm-ostree system

**Comment by Luap99** [comment]
`podman generate systemd` for pods will create a unit for each container inside the pod. Just use it with the `--files` option.
Here are the examples from the man page: https://github.com/containers/podman/blob/master/docs/source/markdown/podman-generate-systemd.1.md#generate-systemd-unit-files-for-a-pod-with-two-simple-alpine-containers

**Comment by barnscott** [comment]
yea, those were the steps i was following.

using --files generates an unpreferable file name, which is why i didnt use that flag.

i dont have any issues with the steps when deploying rootless container, but i get the previously posted errors when using for pod. 

below is the error again. you can see the error is initiated from configs that are generated by podman-generate-systemd:

nextcloud.service: Service has more than one ExecStart= setting, which is only allowed for Type=oneshot services. Refusing.

**Comment by barnscott** [comment]
> `podman generate systemd` for pods will create a unit for each container inside the pod. Just use it with the `--files` option.
> Here are the examples from the man page: https://github.com/containers/podman/blob/master/docs/source/markdown/podman-generate-systemd.1.md#generate-systemd-unit-files-for-a-pod-with-two-simple-alpine-containers

ah, ok. i got it. i simply need to make sure to create a file for each container in the pod. ill try this, thanks!

**Comment by Luap99** [comment]
Just use it with `--files` on a pod. If you want to change the name take a look at the `--pod-prefix`,`--container-prefix` and `--separator` flags.

**Comment by jayache80** [comment]
Sorry to dig this up but I'm hitting a similar issue and need help. I am running rootless, and have created a multi-container pod using `--infra=true` (which allows `podman generate systemd <pod-name>` to work).

What I don't understand is why is a separate unit file generated for *each* container? That's exactly what I don't want. What I want is a *single* unit file for *the pod*. Anyhow, "I'll just use the unit file associated with the pod ID and ignore the other generated ones" I say. And so I do, and the service starts fine (all containers are correctly started). However, if I try to `systemctl --user stop <pod>.service` it doesn't stop the pod nor the containers therein but instead just puts the pod in a "degraded" state. I don't understand because if I manually run the same `ExecStart` and `ExecStop` commands that are in the unit files:

    podman pod start <pod-id>
    podman pod stop <pod-id>

then `podman` starts and stops the pod and all containers as expected. But when systemd tries to do the same, it can't stop the containers and puts the pod in a "degraded" state. What might I be doing wrong?

**Comment by jayache80** [comment]
Sorry again for the noise, and to answer my own question in case my stumbling is helpful to those who might be searching for similar things:

*This is just how systemd works:* it manages each container process separately.

If we want to "cluster" the container services intuitively (as one might expect of a "pod"), such that if you'd like `systemctl --user stop <pod>` to automatically stop *all* containers associated with the pod, then you have to group them using normal systemd conventions. I accomplished this with the help of [this](https://unix.stackexchange.com/questions/440215/stopping-systemd-unit-together-with-another-starting-works).

tl;dr: Essentially, create a "parent" `.target` systemd unit, with `Requires` and `After` referencing all the container `.service` units as well as the pod `.service`. Further, go to each `.service` unit and add a `PartOf=<parent.target>`.

Allow me to illustrate. I have created a pod, called `pod_foo`, consisting of a Python/Gunicorn application, having a container named `foo_application`, and an nginx web server, having a container named `foo_web`. Because I created the pod with `--infra=true`, an additional infrastructure container (an instance of a "pause" image having container name `4a668a810b30-infra`) is created, but just think of that as "being" the pod.

So `podman pod ls` shows:
```
POD ID        NAME        STATUS      CREATED             INFRA ID      # OF CONTAINERS
4a668a810b30  pod_foo     Stopped     About a minute ago  17faddf6fe51  3
```

And `podman container ls -a` shows:
```
CONTAINER ID  IMAGE                             COMMAND               CREATED        STATUS                    PORTS                   NAMES
17faddf6fe51  k8s.gcr.io/pause:3.5                                    3 minutes ago  Created                                           4a668a810b30-infra
4946f59fea88  localhost/foo_application:latest  gunicorn -w 4 --b...  3 minutes ago  Exited (0) 3 minutes ago  0.0.0.0:5000->5000/tcp  foo_application
63a48605b15b  docker.io/library/nginx:latest    nginx -g daemon o...  3 minutes ago  Exited (0) 3 minutes ago  0.0.0.0:8080->80/tcp    foo_web
```

Now I'd like systemd service units to wrap this. So I run:

    podman generate systemd pod_foo --files --name

And it generates the following files:

```
# container-foo_application.service
# autogenerated by Podman 3.4.4
# Sun Feb 27 20:16:50 PST 2022

[Unit]
Description=Podman container-foo_application.service
Documentation=man:podman-generate-systemd(1)
Wants=network-online.target
After=network-online.target
RequiresMountsFor=/run/user/1000/containers

[Service]
Environment=PODMAN_SYSTEMD_UNIT=%n
Restart=on-failure
TimeoutStopSec=70
ExecStart=/usr/bin/podman start foo_application
ExecStop=/usr/bin/podman stop -t 10 foo_application
ExecStopPost=/usr/bin/podman stop -t 10 foo_application
PIDFile=/run/user/1000/containers/overlay-containers/4946f59fea88452a56da36da844d98ce4d9b2c676b588aaa41ca9b7544c7640d/userdata/conmon.pid
Type=forking

[Install]
WantedBy=default.target
```

```
# container-foo_web.service
# autogenerated by Podman 3.4.4
# Sun Feb 27 20:16:50 PST 2022

[Unit]
Description=Podman container-foo_web.service
Documentation=man:podman-generate-systemd(1)
Wants=network-online.target
After=network-online.target
RequiresMountsFor=/run/user/1000/containers

[Service]
Environment=PODMAN_SYSTEMD_UNIT=%n
Restart=on-failure
TimeoutStopSec=70
ExecStart=/usr/bin/podman start foo_web
ExecStop=/usr/bin/podman stop -t 10 foo_web
ExecStopPost=/usr/bin/podman stop -t 10 foo_web
PIDFile=/run/user/1000/containers/overlay-containers/63a48605b15b4ee7d86afc34bf0c61e5255294183d6bca2f59e7dcafe92bc64a/userdata/conmon.pid
Type=forking

[Install]
WantedBy=default.target
```

```
# pod-pod_foo.service
# autogenerated by Podman 3.4.4
# Sun Feb 27 20:16:50 PST 2022

[Unit]
Description=Podman pod-pod_foo.service
Documentation=man:podman-generate-systemd(1)
Wants=network-online.target
After=network-online.target
RequiresMountsFor=
Requires=container-foo_application.service container-foo_web.service
Before=container-foo_application.service container-foo_web.service

[Service]
Environment=PODMAN_SYSTEMD_UNIT=%n
Restart=on-failure
TimeoutStopSec=70
ExecStart=/usr/bin/podman start 4a668a810b30-infra
ExecStop=/usr/bin/podman stop -t 10 4a668a810b30-infra
ExecStopPost=/usr/bin/podman stop -t 10 4a668a810b30-infra
PIDFile=/run/user/1000/containers/overlay-containers/17faddf6fe5165ba3ef68d1c7c0d4d408cab361fe70a7f47f72c0d475476980e/userdata/conmon.pid
Type=forking

[Install]
WantedBy=default.target
```

I can copy those into `~/.config/systemd/user/` and start up the whole deployment simply with:

    systemctl --user start pod-pod_foo.service

and this is fine and good if I don't mind the containers being "independent" of the pod process. However, running:

    systemctl --user stop pod-pod_foo.service

will not stop all the containers. It will only stop the pod "infra" container, leaving the pod in a "degraded" state with all other containers still running.  I'd have to issue a stop to each container to stop them.

However, I can group all the units together using a systemd target. To do that, I created a new file:

```
# foo.target
# Manually created by a human

[Unit]
Description=Web Services for foo
Requires=pod-pod_foo.service container-foo_application.service container-foo_web.service
After=pod_foo.service container-foo_application.service container-foo_web.service

[Install]
WantedBy=default.target
```

This requires going to each of those `.service` files and adding this to the `[Unit]` section:

    PartOf=foo.target

Similar to before, I can start the whole fleet with:

    systemctl --user start foo.target

but now I can issue:

    systemctl --user stop foo.target

and all the containers are stopped in concert.

Whether this approach has any merit, is sustainable, or is maintainable, I'm not sure, and I would appreciate any feedback. I'll make a feature request to see if this `PartOf` business can be optionally autogenerated when `podman-generate-systemd` is working with a pod and see where that goes.
